docs/internals/layout_in_memory.rst

```diff
@@ -1,40 +1,123 @@
 
-Scratch space can be used between statements (i.e. within inline assembly). The zero slot
-is used as initial value for dynamic memory arrays and should never be written to
-(the free memory pointer points to ``0x80`` initially).
+Every time you access memory, the EVM knows the location that was accessed. Roughly speaking, the
+maximum location you access determines the size of active memory. Increasing the size of
+active memory increases your gas costs non-linearly. The specifics are beyond the scope
+of this documentation. More information about gas costs can be found in the Etherum Yellow Paper
+and in the EVM (go-ethereum) project and specifications.
+
+Solidity is a high level language that is translated (compiled) into EVM instructions. Its features
+work because of the conventions Solidity defines and uses, in combination with the implementations
+of complex data structures that it provides (such as multi-dimensional arrays). These data
+structures require that memory be allocated (reserved) for their use, so that instructions that
+execute after they're allocated don't corrupt them.
+
+Memory allocation is a high-level concept, not an EVM concept. Its implementation in Solidity relies
+on the value stored at memory location ``0x40``. That value tracks the next unallocated memory
+address available. This is different from the highest memory address ever accessed by the contract's
+code. Memory access opcodes in inline assembly are not restricted by the value of the free memory pointer and can freely refer to memory located beyond it.
+
+The active memory size
+transaction) is available via the inline assembly instruction ``msize()`` (see
+https://docs.soliditylang.org/en/latest/yul.html#evm-dialect).
+There is no dedicated opcode for accessing the free memory pointer. The address ``0x40`` is
+just an ordinary memory location that is managed by code automatically generated by the compiler. In inline assembly
+memory can be allocated manually by increasing the value of the pointer.
+
+The free memory pointer is central to Solidity's memory allocation implementation,
+but it is effectively independent of the value returned by ``msize()``. Of course, accessing the
+free memory pointer itself (by definition) affects the value returned by ``msize()`` because after
+such an access, the largest memory address accessed would be at least the location of the free
+memory pointer plus its size: ``0x40+0x20``. The free memory pointer is accessed almost immediately
+in any Solidity contract, so this is moot.
+
+The memory between locations ``0x00`` and ``0x3F`` (inclusive) is called "scratch space." It exists
+solely by convention. Solidity's generated code observes the convention that this area is for
+extremely brief temporary use only. Since there is no guarantee that code not under your direct
+control will not change that memory, you should not rely on its value being preserved between statements
+except within the current assembly block.
```

Comment:
This is an interesting addition to the documentation. Probably can even go to the security considerations section.